event_model/__init__.py

```diff
@@ -180,6 +180,103 @@ def bulk_datum(self, doc):
         self.datum_page(bulk_datum_to_datum_page(doc))
 
 
+class SingleRunDocumentRouter(DocumentRouter):
+    """
+    A DocumentRouter intended to process events from exactly
+    one run.
+    """
+    def __init__(self):
+        self._start_doc = None
+        self._descriptors = dict()
+
+    def __call__(self, name, doc, validate=False):
+        """
+        Keep track of the start document and descriptor documents
+        passed to this SingleRunDocumentRouter.
+
+        Send documents to the superclass for processing.
+
+        Parameters
+        ----------
+        name : string
+        doc : dict
+        validate : boolean
+            Apply jsonschema validation to the documents coming *out*. This is
+            False by default.
+
+        Returns
+        -------
+        name, output_doc : string, dict
+            The same name as what was passed in, and a doc that may be the same
+            instance as doc, a copy of doc, or a different dict altogether.
+        """
+        if name == 'start':
+            if self._start_doc is None:
+                self._start_doc = doc
+            else:
+                raise EventModelError(
+                    'SingleRunDocumentRouter received a second start document: uid {doc["uid"]}'
+                )
+        elif name == 'descriptor':
+            if doc['run_start'] == self._start_doc['uid']:
+                self._descriptors[doc['uid']] = doc
+            else:
+                raise EventModelError(
+                    'SingleRunDocumentRouter received a descriptor associated with a different start document'
+                )
+
+        return super().__call__(name, doc, validate)
+
+    def get_start(self):
+        """Convenience method returning the start document for the associated run.
+
+        If no start document has been processed EventModelError will be raised.
+
+        Returns
+        -------
+        start document : dict
+        """
+        if self._start_doc is None:
+            raise EventModelError('SingleRunDocumentRouter has not processed a start document yet')
+
+        return self._start_doc
+
+    def get_descriptor(self, doc):
+        """Convenience method returning the descriptor associated with the specified document.
+
+        Parameters
+        ----------
+        doc : dict
+            event-model document
+
+        Returns
+        -------
+        descriptor document : dict
+        """
+        if 'descriptor' not in doc:
```

Comment:
Micro-optimization: it is faster to "beg forgiveness" than "ask permission".

```
try:
    descriptor_uid = doc['descriptor']
except KeyError:
    ...
try:
    self._descriptors[descriptor_uid]
except KeyError:
    ...
```

Reply:
Nope.

Reply:
A guideline which I picked up from a StackOverflow a long time ago is, "If it's going to work most of the time, begging forgiveness is faster. If it's going to fail often, ask permission." I checked if this guideline is still valid and it looks like it is.

```py
In [1]: d = {'a': 1, 'b': 2, 'c': 3, 'd': 4}                                                                                                                                                  

In [2]: %%timeit 
   ...: try: 
   ...:     d['c'] 
   ...: except KeyError: 
   ...:     ... 
   ...:                                                                                                                                                                                       
43.8 ns ± 1.12 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [3]: %%timeit 
   ...: if 'c' in d: 
   ...:     d['c'] 
   ...:                                                                                                                                                                                       
68.3 ns ± 1.32 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Now, the _fail_ path is much slower...

```py
In [10]: %%timeit 
    ...: if 'x' in d: 
    ...:     d['c'] 
    ...:                                                                                                                                                                                      
39 ns ± 0.989 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [11]: %%timeit 
    ...: try: 
    ...:     d['x'] 
    ...: except KeyError: 
    ...:     ... 
    ...:                                                                                                                                                                                      
252 ns ± 2.95 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

...but in that path the slowness is unimportant because we are raising an unrecoverable exception.

If "Nope" means, "This micro-optimization is not worth bothering with," fair enough. If "Nope" means, "I get different results," then I'm intrigued even if this is just an academic exercise. :- )

Reply:
Sorry, I have been regretting this "Nope" for some time. I should have written up exactly what you have here with my reasoning which is that the work being done with the documents will be significantly more expensive than these checks and so no performance improvement would be realized at the cost of using the wrong language feature. I first saw this optimization in a "biocomputation with python" book and I (still) think if a program is slow because a hot loop is checking for a key in a dictionary the solution is not to replace `in` with `try/catch` but do something smarter than check for a key in the loop.

Apologies for my snippy response and the time you spent writing what I should have written.

Reply:
No worries at all, @jklynch.